engine/pkg/client/internal/leader_resolver.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+	"math/rand"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/engine/pkg/client/internal/endpoint"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+const (
+	schemaName = "tiflow"
+)
+
+// MasterServerList stores a map from server addresses to whether they are the leader.
+type MasterServerList = map[string]bool
+
+// FollowerSorterFn is a function used to customize the order of fail-over
+// in case the servermaster's leader is unreachable.
+type FollowerSorterFn = func(address []resolver.Address)
+
+// LeaderResolver implements a gRPC resolver that handles the
+// follower logic for servermaster clients.
+type LeaderResolver struct {
+	*manual.Resolver
+	serviceConfig *serviceconfig.ParseResult
+
+	serverListMu   sync.RWMutex
+	serverList     MasterServerList
+	updateNotifyCh chan struct{}
+
+	doneCh chan struct{}
+	wg     sync.WaitGroup
+
+	FollowerSorter FollowerSorterFn
```

Comment:
What scenarios we need to use an abstract `FollowerSorter`?
Will we have the scenary that some `server-master` and `executor` are 'closer'?

Reply:
> What scenarios we need to use an abstract `FollowerSorter`? Will we have the scenary that some `server-master` and `executor` are 'closer'?

We need deterministic results in unit tests, but we need randomized results in production.

Reply:
Ok. It's only for easy test, but not an abstract for  multiple ` follower selection strategy`.